src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -515,6 +517,44 @@ public Task SendFunctionEnvironmentReloadRequest()
             return _reloadTask.Task;
         }
 
+        public Task SendWorkerWarmupRequest()
+        {
+            bool capabilityEnabled = !string.IsNullOrEmpty(_workerCapabilities.GetCapabilityState(RpcWorkerConstants.HandlesWorkerWarmupMessage));
+            if (!capabilityEnabled)
+            {
+                _workerChannelLogger.LogDebug("Worker warmup capability not enabled");
+            }
+            else
+            {
+                _workerChannelLogger.LogDebug("Sending WorkerWarmupRequest to WorkerProcess with Pid: '{0}'", _rpcWorkerProcess.Id);
+
+                RegisterCallbackForNextGrpcMessage(MsgType.WorkerWarmupResponse, TimeSpan.FromMinutes(1), 1,
+                msg => ProcessWorkerWarmupResponse(msg.Message.WorkerWarmupResponse), HandleWorkerWarmupError);
+
+                var request = new WorkerWarmupRequest()
+                {
+                    WorkerDirectory = _workerConfig.Description.WorkerDirectory,
+                };
+
+                SendStreamingMessage(new StreamingMessage
+                {
+                    WorkerWarmupRequest = request
+                });
+            }
+            return _workerWarmupTask.Task;
+        }
+
+        internal void ProcessWorkerWarmupResponse(WorkerWarmupResponse response)
+        {
+            _workerChannelLogger.LogDebug("Received WorkerWarmupResponse from WorkerProcess with Pid: '{0}'", _rpcWorkerProcess.Id);
+            if (response.Result.IsFailure(out Exception workerWarmupException))
+            {
+                _workerChannelLogger.LogError(workerWarmupException, "Worker warmup failed");
+                _workerWarmupTask.SetException(workerWarmupException);
+            }
+            _workerWarmupTask.SetResult(true);
```

Comment:
In the case of failure result (expression in line 550 is true), won't line 555 t[hrow an exception since the underlying task is already in the Faulted state](https://learn.microsoft.com/en-us/dotnet/api/System.Threading.Tasks.TaskCompletionSource-1.SetResult?view=net-6.0#exceptions)?

Reply:
updated logic to not use _workerWarmupTask as it is not necessary